### PR TITLE
Professional

### PR DESCRIPTION
**Pull request by TehanovAnton**
Implement professional model

### CODE REVIEW

File: app/models/professional.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class Professional < User
+  default_scope { where(role: :professional) }
+end
```
Comment (dmokreckiy) [inline app/models/professional.rb:3]: End line missing

File: spec/factories/users.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+FactoryBot.define do
+  factory :user, class: 'User' do
+    first_name { Faker::Name.first_name }
+    last_name { Faker::Name.last_name }
+    patronymic { Faker::Name.middle_name }
```
Comment (dmokreckiy) [inline spec/factories/users.rb:5]: that is called `middle_name`

File: spec/factories/users.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    salon_id { rand(1...101) }
```
Comment (dmokreckiy) [inline spec/factories/users.rb:6]: this should either create salon in after create hook, or should be provided on invocation

File: spec/models/professional_spec.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+require 'rails_helper'
+
+RSpec.shared_examples "invalid professional" do |args|
+  subject { FactoryBot.build(:professional, args[:field]=>args[:value]) }
+
+  it { is_expected.to be_invalid }
+end
+
+RSpec.describe Professional, type: :model do
+  describe  'validation' do
```
Comment (dmokreckiy) [inline spec/models/professional_spec.rb:10]: double space. validation`s`

File: spec/models/professional_spec.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    context "presence with nil" do
```
Comment (dmokreckiy) [inline spec/models/professional_spec.rb:11]: find a better naming
context should start with `when xxx` - https://www.betterspecs.org/#contexts

File: spec/models/professional_spec.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+      include_examples 'invalid professional', { field: 'salon_id', value: nil }
+      include_examples 'invalid professional', { field: 'status', value: nil }
+      include_examples 'invalid professional', { field: 'work_email', value: nil }
+      include_examples 'invalid professional', { field: 'work_phone', value: nil }
+      include_examples 'invalid professional', { field: 'rating', value: nil }
+    end
+  
+    context "work email" do
```
Comment (dmokreckiy) [inline spec/models/professional_spec.rb:19]: describe

File: spec/models/professional_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      context "uniqnuess" do
```
Comment (dmokreckiy) [inline spec/models/professional_spec.rb:20]: same as above. uniq`u`ness

File: spec/models/visit_spec.rb [hunk 8]
```diff
@@ -3,31 +3,31 @@
 RSpec.describe Visit, type: :model do
   describe 'does not pass validations with' do
     context 'nil start_date' do
-      subject { FactoryBot.build(:random_visit, :nil_start_at) }
+      subject { build(:random_visit, :nil_start_at) }
```
Comment (dmokreckiy) [inline spec/models/visit_spec.rb:6]: I prefer explicit FactorBot. Otherwise set it as a defined code style within whole team.

**Review by dmokreckiy — CHANGES_REQUESTED** [review]
Please check my comments

File: spec/factories/users.rb (hunk 3 above)
Comment (TehanovAnton) [reply spec/factories/users.rb:6]: Ok, i do it when associations will be ready

**Review by GEK62 — APPROVED** [review]

**Review by Vgoylo — APPROVED** [review]

**Review by Rub1st — APPROVED** [review]

**Comment by Rub1st** [comment]
fix conflicts

File: app/models/user.rb [hunk 9]
```diff
@@ -2,26 +2,48 @@ class User < ApplicationRecord
...
+  validates :middle_name,
+            length: { maximum: 255 },
+            allow_blank: true
+
+  validates :email,
+            uniqueness: { case_sensitive: false },
+            format: { with: URI::MailTo::EMAIL_REGEXP, message: 'Email invalid' },
+            length: { minimum: 4, maximum: 254 }
+
+  validates :phone, format: { with: /(\+375|80) (29|44|33|25) \d{3}-\d{2}-\d{2}/, message: 'Phone invalid' }
```
Comment (TehanovAnton) [inline app/models/user.rb:24]: move regex to constants

**Pull request merged**